In [56]:
#question1
from numpy import mean
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, random_state=2404)

print(len(y_test))
#print(iris)
#print(iris.data.shape) #
#print(X_train[0])
#print(X_test[0])

38


In [57]:
#function to calulate euclidian distance
from math import sqrt

import numpy as np


def distance(a,b):
    return np.sqrt(np.sum((a-b)**2))


In [58]:
#question 3
from collections import Counter
import numpy as np

#knn 1
def nn(X_train,X_test,y_train):

    #print(len(y_test))
    pred_labels = []
    #distances = []
    for test_sample in X_test:
        distances =  [distance(test_sample, train_sample) for train_sample in X_train]
        #stroe all distance for each test
        #distances.append(dist)
        #find distance of minimum distance
        index_min_dist = np.argmin(distances)
        #print(index_min_dist)
        pred_labels.append(y_train[index_min_dist])
        pred_labels_convert_array = np.array(pred_labels)
    #print("y_test ",y_test)
    #print("pred_y ",pred_labels_convert_array )
    
    print("accurancy nn ",np.mean(y_test ==pred_labels_convert_array))
    print("error rate ", 1 - np.mean(y_test ==pred_labels_convert_array))
        #print(distances)
        
    return pred_labels_convert_array
nn(X_train,X_test,y_train)



        

accurancy nn  0.9473684210526315
error rate  0.052631578947368474


array([1, 2, 0, 0, 1, 1, 2, 2, 0, 2, 2, 0, 0, 1, 2, 2, 1, 2, 1, 0, 0, 2,
       1, 2, 2, 0, 2, 0, 2, 1, 0, 1, 2, 0, 2, 1, 0, 0])

In [59]:
#knn 3
def knn(X_train, X_test, y_train, k):
    pred_labels = []
    for test_sample in X_test:
        distances = [distance(test_sample, train_sample) for train_sample in X_train]
        index_distances_k = np.argsort(distances)[:k] #get the first 3 nearest distances in term of index
        #print("list of 3 distances ", index_distances_k)
        nearest_labels = [y_train[i]  for i in index_distances_k] #get the three labels
        #print("list of 3 lables ",nearest_labels)
        majority_vote = Counter(nearest_labels).most_common(1)[0][0] #get tuple(key, number times appear) and store anly the key.
        #print(majority_vote)
        pred_labels.append(majority_vote)
        pred_labels_convert_array3 = np.array(pred_labels)
        


    #print("pred_labels_convert_array ",pred_labels_convert_array3)
    print("accuracy k=3 ",np.mean(y_test ==pred_labels_convert_array3))
    return pred_labels_convert_array3
#nn(X_train,X_test,y_train)
knn(X_train,X_test,y_train,3)

accuracy k=3  0.9473684210526315


array([1, 2, 0, 0, 1, 1, 2, 2, 0, 2, 2, 0, 0, 1, 2, 2, 1, 2, 1, 0, 0, 2,
       1, 2, 2, 0, 2, 0, 2, 1, 0, 1, 2, 0, 2, 1, 0, 0])

In [60]:
knn(X_train,X_test,y_train,3)

accuracy k=3  0.9473684210526315


array([1, 2, 0, 0, 1, 1, 2, 2, 0, 2, 2, 0, 0, 1, 2, 2, 1, 2, 1, 0, 0, 2,
       1, 2, 2, 0, 2, 0, 2, 1, 0, 1, 2, 0, 2, 1, 0, 0])

In [61]:
np.mean(knn(X_train,X_test,y_train,3) == nn(X_train,X_test,y_train))

accuracy k=3  0.9473684210526315
accurancy nn  0.9473684210526315
error rate  0.052631578947368474


np.float64(1.0)

In [62]:
'''

import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from collections import Counter

# Load the Iris dataset
data = load_iris()
X = data.data
y = data.target

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2404)

print(X_test[0])
print(X_train[0])

def euclidean_distance(x1, x2):
    return np.sqrt(np.sum((x1 - x2) ** 2))

# Nearest Neighbour Algorithm
def nearest_neighbour(X_train, y_train, X_test):
    predictions = []
    for test_sample in X_test:
        #print(test_sample)
        distances = [euclidean_distance(test_sample, train_sample) for train_sample in X_train]
        #print("dit",distances)
        nearest_index = np.argmin(distances)
        predictions.append(y_train[nearest_index])
    #print(len(predictions))
    return predictions

# K-Nearest Neighbours Algorithm
def k_nearest_neighbours(X_train, y_train, X_test, k=3):
    predictions = []
    for test_sample in X_test:
        distances = [euclidean_distance(test_sample, train_sample) for train_sample in X_train]
        nearest_indices = np.argsort(distances)[:k]
        nearest_labels = [y_train[i] for i in nearest_indices]
        majority_vote = Counter(nearest_labels).most_common(1)[0][0]
        predictions.append(majority_vote)
    return predictions

#print(euclidean_distance(X_test[0],X_train[0]))

# Evaluate the Nearest Neighbour Algorithm
nn_predictions = nearest_neighbour(X_train, y_train, X_test)
nn_accuracy = np.mean(nn_predictions == y_test)
print(f"Nearest Neighbour Accuracy: {nn_accuracy:.2f}")

# Evaluate the K-Nearest Neighbours Algorithm
k = 3
knn_predictions = k_nearest_neighbours(X_train, y_train, X_test, k=k)
knn_accuracy = np.mean(knn_predictions == y_test)
print(f"K-Nearest Neighbours (k={k}) Accuracy: {knn_accuracy:.2f}")
'''


'\n\nimport numpy as np\nfrom sklearn.datasets import load_iris\nfrom sklearn.model_selection import train_test_split\nfrom collections import Counter\n\n# Load the Iris dataset\ndata = load_iris()\nX = data.data\ny = data.target\n\n# Split the dataset into training and testing sets\nX_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2404)\n\nprint(X_test[0])\nprint(X_train[0])\n\ndef euclidean_distance(x1, x2):\n    return np.sqrt(np.sum((x1 - x2) ** 2))\n\n# Nearest Neighbour Algorithm\ndef nearest_neighbour(X_train, y_train, X_test):\n    predictions = []\n    for test_sample in X_test:\n        #print(test_sample)\n        distances = [euclidean_distance(test_sample, train_sample) for train_sample in X_train]\n        #print("dit",distances)\n        nearest_index = np.argmin(distances)\n        predictions.append(y_train[nearest_index])\n    #print(len(predictions))\n    return predictions\n\n# K-Nearest Neighbours Algorithm\ndef k_nearest_neighbours(X_train, y

In [63]:
'''
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

# Load the Iris dataset
data = load_iris()
X = data.data
y = data.target

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y , random_state=2404)

def euclidean_distance(x1, x2):
    return np.sqrt(np.sum((x1 - x2) ** 2))

def conformity_measure(X_train, y_train, X_test, y_test):
    conformity_scores = []
    for i, test_sample in enumerate(X_test):
        same_class_dist = float('inf')
        diff_class_dist = float('inf')
        
        for j, train_sample in enumerate(X_train):
            dist = distance(test_sample, train_sample)
            #dist = euclidean_distance(test_sample, train_sample)
            
            if y_train[j] == y_test[i]:  # Nearest sample of the same class
                if dist < same_class_dist:
                    same_class_dist = dist
            else:  # Nearest sample of a different class
                if dist < diff_class_dist:
                    diff_class_dist = dist
        
        # Calculate the conformity measure
        if same_class_dist > 0:  # Prevent division by zero
            conformity = diff_class_dist / same_class_dist
        else:
            conformity = float('inf')
        
        conformity_scores.append(conformity)
        print(f"Test sample {i}: Same Class Dist = {same_class_dist:.3f}, "
              f"Different Class Dist = {diff_class_dist:.3f}, "
              f"Conformity = {conformity:.3f}")
    
    return conformity_scores

# Compute conformity measures
#conformity_scores = conformity_measure(X_train, y_train, X_test, y_test)
'''

'\nimport numpy as np\nfrom sklearn.datasets import load_iris\nfrom sklearn.model_selection import train_test_split\n\n# Load the Iris dataset\ndata = load_iris()\nX = data.data\ny = data.target\n\n# Split data into training and testing sets\nX_train, X_test, y_train, y_test = train_test_split(X, y , random_state=2404)\n\ndef euclidean_distance(x1, x2):\n    return np.sqrt(np.sum((x1 - x2) ** 2))\n\ndef conformity_measure(X_train, y_train, X_test, y_test):\n    conformity_scores = []\n    for i, test_sample in enumerate(X_test):\n        same_class_dist = float(\'inf\')\n        diff_class_dist = float(\'inf\')\n        \n        for j, train_sample in enumerate(X_train):\n            dist = distance(test_sample, train_sample)\n            #dist = euclidean_distance(test_sample, train_sample)\n            \n            if y_train[j] == y_test[i]:  # Nearest sample of the same class\n                if dist < same_class_dist:\n                    same_class_dist = dist\n            else

In [64]:
'''

import numpy as np

def euclidean_distance(x1, x2):
    return np.sqrt(np.sum((x1 - x2) ** 2))

def conformity_measure(X_train, y_train, X_test, assumed_labels, epsilon=1e-8):
    conformity_scores = []

    # Repeat assumed_label for all test samples
    assumed_labels = np.full(len(X_test), assumed_labels)
    
    # Extend the training samples and labels
    extended_samples = np.vstack([X_train, X_test])
    extended_labels = np.append(y_train, assumed_labels)

    # Compute conformity scores for the test samples
    for i, sample in enumerate(X_test):  # Only iterate over test samples
        same_class_dist = float('inf')
        diff_class_dist = float('inf')

        for j, train_sample in enumerate(extended_samples):
            dist = euclidean_distance(sample, train_sample)

            if extended_labels[j] == assumed_labels[i]:  # Same class
                same_class_dist = min(same_class_dist, dist)
            else:  # Different class
                diff_class_dist = min(diff_class_dist, dist)

        # Handle division by zero
        if same_class_dist == 0.0:
            same_class_dist += epsilon
        
        conformity = diff_class_dist / same_class_dist
        conformity_scores.append((sample, assumed_labels[i], conformity))

    return conformity_scores


# Example usage with Iris dataset
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

# Load the Iris dataset
data = load_iris()
X = data.data
y = data.target

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2404)

# Evaluate for each assumed label (0, 1, 2)
for label in [0, 1, 2]:
    print(f"\nConformity Scores assuming label {label}:")
    conformity_scores = conformity_measure(X_train, y_train, X_test, assumed_labels=label)
    #print(len(conformity_scores))
    for sample, assumed_label, score in conformity_scores:
        print(f"Test Sample: {sample}, Assumed Label: {assumed_label}, Conformity Score: {score:.3f}")
print(len(conformity_scores))
'''

'\n\nimport numpy as np\n\ndef euclidean_distance(x1, x2):\n    return np.sqrt(np.sum((x1 - x2) ** 2))\n\ndef conformity_measure(X_train, y_train, X_test, assumed_labels, epsilon=1e-8):\n    conformity_scores = []\n\n    # Repeat assumed_label for all test samples\n    assumed_labels = np.full(len(X_test), assumed_labels)\n    \n    # Extend the training samples and labels\n    extended_samples = np.vstack([X_train, X_test])\n    extended_labels = np.append(y_train, assumed_labels)\n\n    # Compute conformity scores for the test samples\n    for i, sample in enumerate(X_test):  # Only iterate over test samples\n        same_class_dist = float(\'inf\')\n        diff_class_dist = float(\'inf\')\n\n        for j, train_sample in enumerate(extended_samples):\n            dist = euclidean_distance(sample, train_sample)\n\n            if extended_labels[j] == assumed_labels[i]:  # Same class\n                same_class_dist = min(same_class_dist, dist)\n            else:  # Different class\n

In [65]:
'''

def conformity_measure(X_train, y_train, X_test, possible_labels=[0, 1, 2]):
    conformity_scores = []

    # Loop through each test sample
    print("test sample ",len(X_test))
    print("train sample ",len(X))
    for i, test_sample in enumerate(X_test):
        scores_for_sample = {}  # Dictionary to store scores for assumed labels
        
        # Assume each possible label and calculate conformity measure
        for assumed_label in possible_labels:
            same_class_dist = float('inf')
            diff_class_dist = float('inf')
            
            for j, train_sample in enumerate(X_train):
                dist = euclidean_distance(test_sample, train_sample)
                
                if y_train[j] == assumed_label:  # Same class
                    same_class_dist = min(same_class_dist, dist)
                else:  # Different class
                    diff_class_dist = min(diff_class_dist, dist)
            
            # Calculate the conformity measure
            if same_class_dist > 0:  # Prevent division by zero
                conformity = diff_class_dist / same_class_dist
            else:
                conformity = float('inf')
            
            # Store the conformity score for this assumed label
            scores_for_sample[assumed_label] = conformity
        
        # Append the scores for this test sample
        conformity_scores.append(scores_for_sample)
        
        # Print the results for debugging
        print(f"Test Sample {i}: {scores_for_sample}")
    
    return conformity_scores



def calculate_p_value(X_train, y_train, X_test, y_test):
    conformity_scores_train = []
    
    # Step 1: Compute conformity scores for training samples
    for i, train_sample in enumerate(X_train):
        score = conformity_measure(X_train, y_train, train_sample, y_train[i], exclude_index=i)
        conformity_scores_train.append(score)
    
    p_values = []

    # Step 2: Compute p-values for test samples
    for i, test_sample in enumerate(X_test):
        test_conformity = conformity_measure(X_train, y_train, test_sample, y_test[i])
        
        # Combine training scores and the test conformity score
        combined_scores = conformity_scores_train + [test_conformity]
        
        # Count scores greater than or equal to the test conformity score
        count = sum(1 for score in combined_scores if score <= test_conformity)
        p_value = count / len(combined_scores)
        
        #p_values.append(p_value)
         
        #print(f"Combined Scores: {combined_scores}")
        #print(f"Test Sample {i}: p-value = {p_value:.3f}")
        
        p_values.append((y_test[i],p_value))
        #print("Number of test samples:", len(X_test))
        #print(len(p_values))
    
    return p_values
onformity_scores = conformity_measure(X_train, y_train, X_test, y_test)
'''


'\n\ndef conformity_measure(X_train, y_train, X_test, possible_labels=[0, 1, 2]):\n    conformity_scores = []\n\n    # Loop through each test sample\n    print("test sample ",len(X_test))\n    print("train sample ",len(X))\n    for i, test_sample in enumerate(X_test):\n        scores_for_sample = {}  # Dictionary to store scores for assumed labels\n        \n        # Assume each possible label and calculate conformity measure\n        for assumed_label in possible_labels:\n            same_class_dist = float(\'inf\')\n            diff_class_dist = float(\'inf\')\n            \n            for j, train_sample in enumerate(X_train):\n                dist = euclidean_distance(test_sample, train_sample)\n                \n                if y_train[j] == assumed_label:  # Same class\n                    same_class_dist = min(same_class_dist, dist)\n                else:  # Different class\n                    diff_class_dist = min(diff_class_dist, dist)\n            \n            # Calcul

In [66]:
'''
def conformity_measure_with_train(X_train, y_train, X_test, possible_labels=[0, 1, 2]):
    conformity_scores = []
    
    # Step 1: Calculate conformity scores for TRAIN samples (use true labels)
    print("Calculating conformity scores for TRAIN samples...")
    for i, train_sample in enumerate(X_train):
        same_class_dist = float('inf')
        diff_class_dist = float('inf')
        
        for j, other_sample in enumerate(X_train):
            if i == j:  # Skip comparison with itself
                continue
            
            dist = euclidean_distance(train_sample, other_sample)
            if y_train[j] == y_train[i]:  # Same class
                same_class_dist = min(same_class_dist, dist)
            else:  # Different class
                diff_class_dist = min(diff_class_dist, dist)
        
        # Prevent division by zero
        if same_class_dist > 0:
            conformity = diff_class_dist / same_class_dist
        else:
            conformity = float('inf')
        
        conformity_scores.append(("Train", i, y_train[i], conformity))
        print(f"Train Sample {i}: Label: {y_train[i]}, Conformity Score: {conformity:.3f}")
    
    # Step 2: Calculate conformity scores for TEST samples with assumed labels
    print("\nCalculating conformity scores for TEST samples...")
    for i, test_sample in enumerate(X_test):
        scores_for_sample = {}  # Dictionary to store scores for assumed labels
        
        for assumed_label in possible_labels:
            same_class_dist = float('inf')
            diff_class_dist = float('inf')
            
            for j, train_sample in enumerate(X_train):
                dist = euclidean_distance(test_sample, train_sample)
                if y_train[j] == assumed_label:  # Same class
                    same_class_dist = min(same_class_dist, dist)
                else:  # Different class
                    diff_class_dist = min(diff_class_dist, dist)
            
            # Prevent division by zero
            if same_class_dist > 0:
                conformity = diff_class_dist / same_class_dist
            else:
                conformity = float('inf')
            
            scores_for_sample[assumed_label] = conformity
        
        # Append conformity scores for this test sample
        conformity_scores.append(("Test", i, scores_for_sample))
        print(f"Test Sample {i}: {scores_for_sample}")
    
    return conformity_scores

onformity_scores = conformity_measure_with_train(X_train, y_train, X_test, y_test)
'''

'\ndef conformity_measure_with_train(X_train, y_train, X_test, possible_labels=[0, 1, 2]):\n    conformity_scores = []\n    \n    # Step 1: Calculate conformity scores for TRAIN samples (use true labels)\n    print("Calculating conformity scores for TRAIN samples...")\n    for i, train_sample in enumerate(X_train):\n        same_class_dist = float(\'inf\')\n        diff_class_dist = float(\'inf\')\n        \n        for j, other_sample in enumerate(X_train):\n            if i == j:  # Skip comparison with itself\n                continue\n            \n            dist = euclidean_distance(train_sample, other_sample)\n            if y_train[j] == y_train[i]:  # Same class\n                same_class_dist = min(same_class_dist, dist)\n            else:  # Different class\n                diff_class_dist = min(diff_class_dist, dist)\n        \n        # Prevent division by zero\n        if same_class_dist > 0:\n            conformity = diff_class_dist / same_class_dist\n        else:\n  

In [67]:
'''
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

# Load the Iris dataset
data = load_iris()
X = data.data
y = data.target

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

def euclidean_distance(x1, x2):
    return np.sqrt(np.sum((x1 - x2) ** 2))

def conformity_measure(X_train, y_train, sample, label, exclude_index=None):
    same_class_dist = float('inf')
    diff_class_dist = float('inf')

    for i, train_sample in enumerate(X_train):
        # Skip self-comparison if exclude_index is provided
        if exclude_index is not None and i == exclude_index:
            continue

        dist = np.sqrt(np.sum((sample - train_sample) ** 2))  # Euclidean distance

        if y_train[i] == label:  # Same class
            same_class_dist = min(same_class_dist, dist)
        else:  # Different class
            diff_class_dist = min(diff_class_dist, dist)

    # Handle division by zero properly
    if same_class_dist > 0:
        return diff_class_dist / same_class_dist
    else:
        return float('inf')


def calculate_p_value(X_train, y_train, X_test, y_test):
    conformity_scores_train = []
    
    # Step 1: Compute conformity scores for training samples
    for i, train_sample in enumerate(X_train):
        score = conformity_measure(X_train, y_train, train_sample, y_train[i], exclude_index=i)
        conformity_scores_train.append(score)
    
    p_values = []

    # Step 2: Compute p-values for test samples
    for i, test_sample in enumerate(X_test):
        test_conformity = conformity_measure(X_train, y_train, test_sample, y_test[i])
        
        # Combine training scores and the test conformity score
        combined_scores = conformity_scores_train + [test_conformity]
        
        # Count scores greater than or equal to the test conformity score
        count = sum(1 for score in combined_scores if score <= test_conformity)
        p_value = count / len(combined_scores)
        
        #p_values.append(p_value)
         
        #print(f"Combined Scores: {combined_scores}")
        #print(f"Test Sample {i}: p-value = {p_value:.3f}")
        
        p_values.append((y_test[i],p_value))
        #print("Number of test samples:", len(X_test))
        #print(len(p_values))
    
    return p_values


# Calculate p-values for test samples
p_values = calculate_p_value(X_train, y_train, X_test, y_test)
print(p_values)
print(p_values[1][0])
print(p_values[0][0] == (y_test[0]))

#average error p_value
def average_error_p_value(p_values):

    return p_values

ave = average_error_p_value(p_values)

print("average ",)
'''

'\nimport numpy as np\nfrom sklearn.datasets import load_iris\nfrom sklearn.model_selection import train_test_split\n\n# Load the Iris dataset\ndata = load_iris()\nX = data.data\ny = data.target\n\n# Split data into training and testing sets\nX_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)\n\ndef euclidean_distance(x1, x2):\n    return np.sqrt(np.sum((x1 - x2) ** 2))\n\ndef conformity_measure(X_train, y_train, sample, label, exclude_index=None):\n    same_class_dist = float(\'inf\')\n    diff_class_dist = float(\'inf\')\n\n    for i, train_sample in enumerate(X_train):\n        # Skip self-comparison if exclude_index is provided\n        if exclude_index is not None and i == exclude_index:\n            continue\n\n        dist = np.sqrt(np.sum((sample - train_sample) ** 2))  # Euclidean distance\n\n        if y_train[i] == label:  # Same class\n            same_class_dist = min(same_class_dist, dist)\n        else:  # Different class\n         

In [68]:
def compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label, epsilon=1e-8):
    results = []

    # Add the test sample with assumed label to the training set temporarily
    extended_samples = np.vstack([train_samples, test_sample])
    extended_labels = np.append(train_labels, assumed_label)

    # Compute conformity scores for each sample in the extended training set
    for i, sample in enumerate(extended_samples):
        same_class_dist = float('inf')
        diff_class_dist = float('inf')

        for j, other_sample in enumerate(extended_samples):
            if i == j:  # Skip comparison with itself
                continue
            
            dist = distance(sample, other_sample)

            if extended_labels[j] == extended_labels[i]:  # Same class
                same_class_dist = min(same_class_dist, dist)
            else:  # Different class
                diff_class_dist = min(diff_class_dist, dist)

        # Handle division by zero
        if same_class_dist == 0.0:
            same_class_dist += epsilon
        
        conformity = diff_class_dist / same_class_dist
        results.append((sample, extended_labels[i], conformity))
    
    return results

In [69]:
#calulate pValue for assumed 0
def pValue(test_score_neg, confScore0,key):
    list_pValue_with_lable0  = []

    rank = 0
    for score in confScore0:
        if score <= test_score_neg:
            rank +=1
        else:
            continue
    
    p_value = rank / len(confScore0)
    list_pValue_with_lable0.append((key,p_value))
    #print("list pValue0\n",list_pValue_with_lable0)
    return list_pValue_with_lable0 

In [70]:
'''
def list_score_test_sample(conformity_scores_zero,key):
    seeZero = []
    zeroF = []
    list_score_test_sample_0 = []
    count = 0
    for sample, label, score in conformity_scores_zero:
        if(count <= 105):
            seeZero.append(score)
            count += 1
        else:
            count = 0
        #print(f"Sample: {sample}, Label: {label}, Conformity Score: {score:.3f}")
    zeroF.append(seeZero.copy())
    seeZero.clear()
    #retrive all measure score for all test sample with assumed label 0
    score_test_sample_0 = conformity_scores_zero[-1][2]
    list_score_test_sample_0.append((key,score_test_sample_0))
    list_score_test_sample_0_array = np.array(list_score_test_sample_0)
    return list_score_test_sample_0_array, zeroF
    '''

'\ndef list_score_test_sample(conformity_scores_zero,key):\n    seeZero = []\n    zeroF = []\n    list_score_test_sample_0 = []\n    count = 0\n    for sample, label, score in conformity_scores_zero:\n        if(count <= 105):\n            seeZero.append(score)\n            count += 1\n        else:\n            count = 0\n        #print(f"Sample: {sample}, Label: {label}, Conformity Score: {score:.3f}")\n    zeroF.append(seeZero.copy())\n    seeZero.clear()\n    #retrive all measure score for all test sample with assumed label 0\n    score_test_sample_0 = conformity_scores_zero[-1][2]\n    list_score_test_sample_0.append((key,score_test_sample_0))\n    list_score_test_sample_0_array = np.array(list_score_test_sample_0)\n    return list_score_test_sample_0_array, zeroF\n    '

In [71]:
import numpy as np

def euclidean_distance(x1, x2):
    return np.sqrt(np.sum((x1 - x2) ** 2))

def compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label, epsilon=1e-8):
    results = []

    # Add the test sample with assumed label to the training set temporarily
    extended_samples = np.vstack([train_samples, test_sample])
    extended_labels = np.append(train_labels, assumed_label)

    # Compute conformity scores for each sample in the extended training set
    for i, sample in enumerate(extended_samples):
        same_class_dist = float('inf')
        diff_class_dist = float('inf')

        for j, other_sample in enumerate(extended_samples):
            if i == j:  # Skip comparison with itself
                continue
            
            dist = euclidean_distance(sample, other_sample)

            if extended_labels[j] == extended_labels[i]:  # Same class
                same_class_dist = min(same_class_dist, dist)
            else:  # Different class
                diff_class_dist = min(diff_class_dist, dist)

        # Handle division by zero
        if same_class_dist == 0.0:
            same_class_dist += epsilon
        
        conformity = diff_class_dist / same_class_dist
        results.append((sample, extended_labels[i], conformity))
    
    return results


#retrive the conformity score of assumed label +1
def retrieve_conformity_scorePos(train_samples, train_labels, test_sample):
    # Compute conformity scores assuming label +1
    conformity_scores_pos = compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label=1)
    test_score_pos = conformity_scores_pos[-1][2]  # Extract conformity score of the test sample
    #print(f"Conformity Score assuming label +1: {test_score_pos:.3f}")
    return test_score_pos

#retrive the conformity score of assumed label -1   
def retrieve_conformity_scoreNeg(train_samples, train_labels, test_sample):
    # Compute conformity scores assuming label -1
    conformity_scores_neg = compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label=-1)
    test_score_neg = conformity_scores_neg[-1][2]  # Extract conformity score of the test sample
    #print(f"Conformity Score assuming label -1: {test_score_neg:.3f}")
    return test_score_neg


#calulate pValue for assumed -1
# Define training data
train_samples = np.array([[0, 3], [2, 2], [3, 3], [-1, 1], [-1, -1], [0, 1]])
train_labels = np.array([1, 1, 1, -1, -1, -1])  # 1 = positive, -1 = negative
#train_samples = X_train
#train_labels = y_train
# Define test sample
test_sample = np.array([0, 0])

# Compute conformity scores assuming label +1
print("Conformity Scores assuming the test sample is +1:")
conformity_scores_zero = compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label=1)
for sample, label, score in conformity_scores_zero:
    print(f"Sample: {sample}, Label: {label}, Conformity Score: {score}")

# Compute conformity scores assuming label -1
print("\nConformity Scores assuming the test sample is -1:")
conformity_scores_neg = compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label=-1)
for sample, label, score in conformity_scores_neg:
    print(f"Sample: {sample}, Label: {label}, Conformity Score: {score}")



confScorePos = retrieve_conformity_scorePos(train_samples, train_labels, test_sample)
print(confScorePos)
confScoreNeg = retrieve_conformity_scoreNeg(train_samples, train_labels, test_sample)
print(f"{confScoreNeg}")


#calulate pValue for assumed +1
def pValuePos(test_score_pos, conformity_scores_pos):
    rank = 0
    for sample, label, score in conformity_scores_pos:
        if score <= test_score_pos:
            rank +=1
        else:
            continue
    
    p_value = rank / len(conformity_scores_pos) 
    
    return p_value 


#calulate pValue for assumed -1
def pValueNeg(test_score_neg, conformity_scores_neg):
    rank = 0
    for sample, label, score in conformity_scores_neg:
        if score <= test_score_neg:
            rank +=1
        else:
            continue
    
    p_value = rank / len(conformity_scores_neg) 
    #print(len(conformity_scores_neg))
    return p_value 


#calulate pValue for assumed +1
pValueP = pValuePos(confScorePos, conformity_scores_zero)
print(f"p_vlaure for assumed +1 {pValueP:.3f}")

#calulate pValue for assumed -1
pValueN = pValueNeg(confScoreNeg, conformity_scores_neg)
print(f"p_vlaure for assumed -1 {pValueN:.3f}")

Conformity Scores assuming the test sample is +1:
Sample: [0 3], Label: 1, Conformity Score: 0.8944271909999159
Sample: [2 2], Label: 1, Conformity Score: 1.5811388300841895
Sample: [3 3], Label: 1, Conformity Score: 2.5495097567963922
Sample: [-1  1], Label: -1, Conformity Score: 1.4142135623730951
Sample: [-1 -1], Label: -1, Conformity Score: 0.7071067811865476
Sample: [0 1], Label: -1, Conformity Score: 1.0
Sample: [0 0], Label: 1, Conformity Score: 0.35355339059327373

Conformity Scores assuming the test sample is -1:
Sample: [0 3], Label: 1, Conformity Score: 0.8944271909999159
Sample: [2 2], Label: 1, Conformity Score: 1.5811388300841895
Sample: [3 3], Label: 1, Conformity Score: 2.5495097567963922
Sample: [-1  1], Label: -1, Conformity Score: 2.23606797749979
Sample: [-1 -1], Label: -1, Conformity Score: 2.91547594742265
Sample: [0 1], Label: -1, Conformity Score: 2.0
Sample: [0 0], Label: -1, Conformity Score: 2.8284271247461903
0.35355339059327373
2.8284271247461903
p_vlaure f

Conformity Scores assuming the test sample is +1:
Sample: [0 3], Label: 1, Conformity Score: 0.8944271909999159
Sample: [2 2], Label: 1, Conformity Score: 1.5811388300841895
Sample: [3 3], Label: 1, Conformity Score: 2.5495097567963922
Sample: [-1  1], Label: -1, Conformity Score: 1.4142135623730951
Sample: [-1 -1], Label: -1, Conformity Score: 0.7071067811865476
Sample: [0 1], Label: -1, Conformity Score: 1.0
Sample: [0 0], Label: 1, Conformity Score: 0.35355339059327373

Conformity Scores assuming the test sample is -1:
Sample: [0 3], Label: 1, Conformity Score: 0.8944271909999159
Sample: [2 2], Label: 1, Conformity Score: 1.5811388300841895
Sample: [3 3], Label: 1, Conformity Score: 2.5495097567963922
Sample: [-1  1], Label: -1, Conformity Score: 2.23606797749979
Sample: [-1 -1], Label: -1, Conformity Score: 2.91547594742265
Sample: [0 1], Label: -1, Conformity Score: 2.0
Sample: [0 0], Label: -1, Conformity Score: 2.8284271247461903
0.35355339059327373
2.8284271247461903
p_vlaure for assumed +1 0.143
p_vlaure for assumed -1 0.857

In [72]:
import numpy as np
'''
def euclidean_distance(x1, x2):
    return np.sqrt(np.sum((x1 - x2) ** 2))

def compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label, epsilon=1e-8):
    results = []

    # Add the test sample with assumed label to the training set temporarily
    extended_samples = np.vstack([train_samples, test_sample])
    extended_labels = np.append(train_labels, assumed_label)

    # Compute conformity scores for each sample in the extended training set
    for i, sample in enumerate(extended_samples):
        same_class_dist = float('inf')
        diff_class_dist = float('inf')

        for j, other_sample in enumerate(extended_samples):
            if i == j:  # Skip comparison with itself
                continue
            
            dist = distance(sample, other_sample)

            if extended_labels[j] == extended_labels[i]:  # Same class
                same_class_dist = min(same_class_dist, dist)
            else:  # Different class
                diff_class_dist = min(diff_class_dist, dist)

        # Handle division by zero
        if same_class_dist == 0.0:
            same_class_dist += epsilon
        
        conformity = diff_class_dist / same_class_dist
        results.append((sample, extended_labels[i], conformity))
    
    return results
'''
# Define training data
#train_samples = np.array([[0, 3], [2, 2], [3, 3], [-1, 1], [-1, -1], [0, 1]])
#train_labels = np.array([1, 1, 1, -1, -1, -1])  # 1 = positive, -1 = negative
list_score_test_sample_0 = []
list_score_test_sample_1 = []
list_score_test_sample_2 = []
#count = 0
seeZero = []
zeroF = []
seeOne = []
oneF = []
seeTwo = []
twoF = []
list_pValue0 = []
list_pValue1 = []
list_pValue2 = []
tot_list_pValues = []
tot_list_pValues2 = []
for i in X_test: 
    train_samples = X_train
    train_labels = y_train
    #print("y_train ",len(y_train))
    # Define test sample
    #test_sample = np.array([0, 0])

    test_sample = i

    # Compute conformity scores assuming label 0    
    #print("\nConformity Scores assuming the test sample is 0:")
    conformity_scores_zero = compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label=0)
 
    count = 0
    for sample, label, score in conformity_scores_zero:
        if(count <= 105):
            seeZero.append(score)
            count += 1
        else:
            count = 0
        #print(f"Sample: {sample}, Label: {label}, Conformity Score: {score:.3f}")
    zeroF.append(seeZero.copy())
    seeZero.clear()
    #retrive all measure score for all test sample with assumed label 0
    score_test_sample_0 = conformity_scores_zero[-1][2]
    list_score_test_sample_0.append((0,score_test_sample_0))
    list_score_test_sample_0_array = np.array(list_score_test_sample_0)


    # Compute conformity scores assuming label 1
    #print("\nConformity Scores assuming the test sample is 1:")
    conformity_scores_one = compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label=1)

    count = 0
    for sample, label, score in conformity_scores_one:
        if(count <= 105):
            seeOne.append(score)
            count += 1
        else:
            count = 0
        #print(f"Sample: {sample}, Label: {label}, Conformity Score: {score:.3f}")
    oneF.append(seeOne.copy())
    seeOne.clear()
    #print(f"Sample: {sample}, Label: {label}, Conformity Score: {score:.3f}")
    #retrive all measure score for all test sample with assumed label 1
    score_test_sample_1 = conformity_scores_one[-1][2]
    list_score_test_sample_1.append((1,score_test_sample_1))
    list_score_test_sample_1_array = np.array(list_score_test_sample_1)

    # Compute conformity scores assuming label 2
    #print("\nConformity Scores assuming the test sample is 2:")
    conformity_scores_two = compute_conformity_scores(train_samples, train_labels, test_sample, assumed_label=2)
    count = 0
    for sample, label, score in conformity_scores_two:
        if(count <= 105):
            seeTwo.append(score)
            count += 1
        else:
            count = 0
        #print(f"Sample: {sample}, Label: {label}, Conformity Score: {score:.3f}")
    twoF.append(seeTwo.copy())
    seeTwo.clear()
    #print(f"Sample: {sample}, Label: {label}, Conformity Score: {score:.3f}")
    
     #retrive all measure score for all test sample with assumed label 2
    score_test_sample_2 = conformity_scores_two[-1][2]
    list_score_test_sample_2.append((2,score_test_sample_2))
    list_score_test_sample_2_array = np.array(list_score_test_sample_2)


np.set_printoptions(suppress=True, precision=40)  # Suppress scientific notation, 5 decimal places
#print(f"list assumed label 0\n{list_score_test_sample_0_array}")
#print(f"list assumed label 1\n{list_score_test_sample_1_array}")
#print(f"list assumed label 2\n{list_score_test_sample_2_array}")


'''
#calulate pValue for assumed 0
def pValueZero(test_score_neg, confScore0):
    list_pValue_with_lable0  = []

    rank = 0
    for score in confScore0:
        if score <= test_score_neg:
            rank +=1
        else:
            continue
    
    p_value = rank / len(confScore0)
    list_pValue_with_lable0.append((0,p_value))
    #print("list pValue0\n",list_pValue_with_lable0)
    return list_pValue_with_lable0 

#calulate pValue for assumed +1
def pValueOne(test_score_neg, confScore1):
    list_pValue_with_lable1  = []

    rank = 0
    for score in confScore1:
        if score <= test_score_neg:
            rank +=1
        else:
            continue
    
    p_value = rank / len(confScore1)
    list_pValue_with_lable1.append((1,p_value))
    #print(len(confScore1))
    return list_pValue_with_lable1 

#calulate pValue for assumed +2
def pValueTwo(test_score_neg, confScore2):
    list_pValue_with_lable2  = []

    rank = 0
    for score in confScore2:
        if score <= test_score_neg:
            rank +=1
        else:
            continue
    
    p_value = rank / len(confScore2)
    list_pValue_with_lable2.append((2,p_value))
    #print("list pValue0\n",list_pValue_with_lable2)
    return list_pValue_with_lable2 
'''
#calulate pValue for assumed 0
for i in range(len(list_score_test_sample_0_array)):
    test_score_0 = list_score_test_sample_0_array[i][1]  # Extract the score from (label, score)
    #print("ciao ",test_score_0)
    conformity_score0 = zeroF[i]  # Get the corresponding list of scores from zeroF
    
    pValue0 = pValue(test_score_0, conformity_score0,key=0)
    list_pValue0.append(pValue0)
    #print(f"p-value for assumed 0 at iteration {pValue0}")

#calulate pValue for assumed 1
for i in range(len(list_score_test_sample_0_array)):
    test_score_One = list_score_test_sample_1_array[i][1]  # Extract the score from (label, score)
    #print("ciao ",test_score_One)
    conformity_score1 = oneF[i]  # Get the corresponding list of scores from zeroF
    
    pValue1 = pValue(test_score_One, conformity_score1,key=1)
    list_pValue1.append(pValue1)

    #print(f"p-value for assumed 1 at iteration {pValue1}")

#calulate pValue for assumed 2
for i in range(len(list_score_test_sample_2_array)):
    test_score_Two = list_score_test_sample_2_array[i][1]  # Extract the score from (label, score)
    #print("ciao ",test_score_Two)
    conformity_score2 = twoF[i]  # Get the corresponding list of scores from zeroF
    
    pValue2 = pValue(test_score_Two, conformity_score2,key=2)
    list_pValue2.append(pValue2)

    #print(f"p-value for assumed 2 at iteration {pValue2}")

for i in range(len(list_pValue0)):
        tot_list_pValues2.append(list_pValue0[i])
        tot_list_pValues2.append(list_pValue1[i])
        tot_list_pValues2.append(list_pValue2[i])

convert_array = np.array(tot_list_pValues2)

print("convert_array\n", convert_array)
'''
print("list pValue0\n",len(list_pValue0))
print("list pValue1\n",list_pValue1)
print("list pValue2\n",list_pValue2)
tot_list_pValues.append(list_pValue0)
tot_list_pValues.append(list_pValue1)
tot_list_pValues.append(list_pValue2)
print("total list pValues\n",tot_list_pValues)



#calulate pValue for assumed 1
for i in range(len(list_score_test_sample_0_array)):
    pValueP = pValueZero(list_score_test_sample_0_array[i], zeroF[i])
    print(f"p_vlaure for assumed 0 and all iteration \n {pValueP} {i}")
'''
#print(f"sees \n {zeroF[0]}")

convert_array
 [[[0.                   0.008928571428571428]]

 [[1.                   0.05357142857142857 ]]

 [[2.                   0.017857142857142856]]

 [[0.                   0.008928571428571428]]

 [[1.                   0.008928571428571428]]

 [[2.                   0.1875              ]]

 [[0.                   0.8214285714285714  ]]

 [[1.                   0.008928571428571428]]

 [[2.                   0.008928571428571428]]

 [[0.                   0.9196428571428571  ]]

 [[1.                   0.008928571428571428]]

 [[2.                   0.008928571428571428]]

 [[0.                   0.008928571428571428]]

 [[1.                   0.5178571428571429  ]]

 [[2.                   0.008928571428571428]]

 [[0.                   0.008928571428571428]]

 [[1.                   0.7857142857142857  ]]

 [[2.                   0.008928571428571428]]

 [[0.                   0.008928571428571428]]

 [[1.                   0.008928571428571428]]

 [[2.                   0

'\nprint("list pValue0\n",len(list_pValue0))\nprint("list pValue1\n",list_pValue1)\nprint("list pValue2\n",list_pValue2)\ntot_list_pValues.append(list_pValue0)\ntot_list_pValues.append(list_pValue1)\ntot_list_pValues.append(list_pValue2)\nprint("total list pValues\n",tot_list_pValues)\n\n\n\n#calulate pValue for assumed 1\nfor i in range(len(list_score_test_sample_0_array)):\n    pValueP = pValueZero(list_score_test_sample_0_array[i], zeroF[i])\n    print(f"p_vlaure for assumed 0 and all iteration \n {pValueP} {i}")\n'

In [73]:
'''
for i in range(len(list_pValue0)):
        #print(list_pValue0[i][0][0])
        #print(list_pValue0[i][0][1])        
'''
sum = 0
count = 0
for i in  range(len(y_test)):
        #print(label,list_pValue0[i][0][0])
        if(y_test[i] == 0):
                #print("first ",list_pValue0[i][0][0],y_test[i])
                sum += list_pValue1[i][0][1] + list_pValue2[i][0][1] 
                count += 2
                #print("counter first", count)
        elif(y_test[i]== 1):
                #print("second ",list_pValue1[i][0][0],y_test[i])
                sum += list_pValue0[i][0][1] + list_pValue2[i][0][1]
                count += 2
                #print("counter secornd", count)
        elif(y_test[i] == 2):
                #print("third ",list_pValue2[i][0][0],y_test[i])
                sum += list_pValue0[i][0][1] + list_pValue1[i][0][1]
                count += 2
                #print("counter third", count)

#print(sum)         
#print("count ",count)
tot = sum /count
print("average false p_value ", tot)



average false p_value  0.012335526315789479


In [74]:
'''
sum_p_value = 0
count = 0

for i in range(len(y_test)):
    if y_test[i] == 0:
        # Sum the p_values from list_pValue1 and list_pValue2
        sum_p_value += list_pValue1[i][0][1] + list_pValue2[i][0][1]
        count += 2  # Increment count for two values
    elif y_test[i] == 1:
        # Sum the p_values from list_pValue0 and list_pValue2
        sum_p_value += list_pValue0[i][0][1] + list_pValue2[i][0][1]
        count += 2
    elif y_test[i] == 2:
        # Sum the p_values from list_pValue0 and list_pValue1
        sum_p_value += list_pValue0[i][0][1] + list_pValue1[i][0][1]
        count += 2

# Calculate and print the average p_value
if count > 0:
    average_p_value = sum_p_value / count
    print("Average false p_value:", average_p_value)
else:
    print("No values to calculate.")

print("Total p_value sum:", sum_p_value)
print("Count of p_values included:", count)
'''

'\nsum_p_value = 0\ncount = 0\n\nfor i in range(len(y_test)):\n    if y_test[i] == 0:\n        # Sum the p_values from list_pValue1 and list_pValue2\n        sum_p_value += list_pValue1[i][0][1] + list_pValue2[i][0][1]\n        count += 2  # Increment count for two values\n    elif y_test[i] == 1:\n        # Sum the p_values from list_pValue0 and list_pValue2\n        sum_p_value += list_pValue0[i][0][1] + list_pValue2[i][0][1]\n        count += 2\n    elif y_test[i] == 2:\n        # Sum the p_values from list_pValue0 and list_pValue1\n        sum_p_value += list_pValue0[i][0][1] + list_pValue1[i][0][1]\n        count += 2\n\n# Calculate and print the average p_value\nif count > 0:\n    average_p_value = sum_p_value / count\n    print("Average false p_value:", average_p_value)\nelse:\n    print("No values to calculate.")\n\nprint("Total p_value sum:", sum_p_value)\nprint("Count of p_values included:", count)\n'